Add api for changing function executor settings, hook up execution with decomposition registry

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #74348
* #74076
* __->__ #74186
* #74012
* #73938
* #73876
* #73875

Make the execution settings mutable on function_impl so that we can set it for running op decompositions. Add mapping to function objects and show example in test of executing op decompositions.

Differential Revision: [D34938125](https://our.internmc.facebook.com/intern/diff/D34938125)

Commit:
Add api for changing function executor settings, hook up execution with decomposition registry

[ghstack-poisoned]

davidberard98 commented on torch/csrc/jit/runtime/decomposition_registry.cpp:
```diff
@@ -108,5 +100,32 @@ void RunDecompositions(std::shared_ptr<Graph> g) {
+
+c10::optional<GraphFunction*> GetDecompositionFunction(
+    const FunctionSchema& schema) {
+  loadDecompositionFunctions();
+  auto cache_it = schema_to_function.find(&schema);
+  GRAPH_DEBUG("Trying to find schema: ", schema);
+  if (cache_it == schema_to_function.end()) {
+    GRAPH_DEBUG("Could not find schema: ", schema);
+    return c10::nullopt;
+  }
```
nit: could you call DecompositionGraphForSchema() here to avoid duplicating this code?

davidberard98 commented on test/cpp/jit/test_misc.cpp:
```diff
@@ -2951,6 +2952,27 @@ graph(%x.1 : Tensor):
+  GraphFunction* func;
+  std::once_flag flag1;
+  for (bool unbiased : {true, false}) {
+    std::call_once(flag1, [&]() {
```
just curious, why std::call_once and not just put this before the loop?

eellison replied:
This is mirroring how I expect the intended use case to go within functorch. You dont want to do the lookup from schema -> function and setting of execution on each call you want to run that code only once.

davidberard98 commented on torch/csrc/jit/runtime/decomposition_registry.cpp:
```diff
@@ -108,5 +100,32 @@ void RunDecompositions(std::shared_ptr<Graph> g) {
+    return c10::nullopt;
+  }
+  auto& func = toGraphFunction(*cache_it->second);
+  func._set_initial_executor_execution_mode(ExecutorExecutionMode::SIMPLE);
```
I guess this is to avoid fusing the ops that were just decomposed? Could you add a comment?

davidberard98 reviewed (approved):
had a few questions/comments, otherwise looks good!

Commit:
Update on "Add api for changing function executor settings, hook up execution with decomposition registry"


Make the execution settings mutable on function_impl so that we can set it for running op decompositions. Add mapping to function objects and show example in test of executing op decompositions.

[ghstack-poisoned]

Commit:
Update on "Add api for changing function executor settings, hook up execution with decomposition registry"


Make the execution settings mutable on function_impl so that we can set it for running op decompositions. Add mapping to function objects and show example in test of executing op decompositions.

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938125).

Commit:
Update on "Add api for changing function executor settings, hook up execution with decomposition registry"


Make the execution settings mutable on function_impl so that we can set it for running op decompositions. Add mapping to function objects and show example in test of executing op decompositions.

Differential Revision: [D34938125](https://our.internmc.facebook.com/intern/diff/D34938125)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938125).

Commit:
Update on "Add api for changing function executor settings, hook up execution with decomposition registry"


Make the execution settings mutable on function_impl so that we can set it for running op decompositions. Add mapping to function objects and show example in test of executing op decompositions.

Differential Revision: [D34938125](https://our.internmc.facebook.com/intern/diff/D34938125)

[ghstack-poisoned]

Commit:
Update on "Add api for changing function executor settings, hook up execution with decomposition registry"


Make the execution settings mutable on function_impl so that we can set it for running op decompositions. Add mapping to function objects and show example in test of executing op decompositions.

Differential Revision: [D34938125](https://our.internmc.facebook.com/intern/diff/D34938125)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938125).

Commit:
Update on "Add api for changing function executor settings, hook up execution with decomposition registry"


Make the execution settings mutable on function_impl so that we can set it for running op decompositions. Add mapping to function objects and show example in test of executing op decompositions.

Differential Revision: [D34938125](https://our.internmc.facebook.com/intern/diff/D34938125)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938125).

Commit:
Update on "Add api for changing function executor settings, hook up execution with decomposition registry"


Make the execution settings mutable on function_impl so that we can set it for running op decompositions. Add mapping to function objects and show example in test of executing op decompositions.

Differential Revision: [D34938125](https://our.internmc.facebook.com/intern/diff/D34938125)

[ghstack-poisoned]

Commit:
Update on "Add api for changing function executor settings, hook up execution with decomposition registry"


Make the execution settings mutable on function_impl so that we can set it for running op decompositions. Add mapping to function objects and show example in test of executing op decompositions.

Differential Revision: [D34938125](https://our.internmc.facebook.com/intern/diff/D34938125)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938125).

Commit:
Update on "Add api for changing function executor settings, hook up execution with decomposition registry"


Make the execution settings mutable on function_impl so that we can set it for running op decompositions. Add mapping to function objects and show example in test of executing op decompositions.

Differential Revision: [D34938125](https://our.internmc.facebook.com/intern/diff/D34938125)

[ghstack-poisoned]

Commit:
Update on "Add api for changing function executor settings, hook up execution with decomposition registry"


Make the execution settings mutable on function_impl so that we can set it for running op decompositions. Add mapping to function objects and show example in test of executing op decompositions.

Differential Revision: [D34938125](https://our.internmc.facebook.com/intern/diff/D34938125)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938125).

Commit:
Update on "Add api for changing function executor settings, hook up execution with decomposition registry"


Make the execution settings mutable on function_impl so that we can set it for running op decompositions. Add mapping to function objects and show example in test of executing op decompositions.

Differential Revision: [D34938125](https://our.internmc.facebook.com/intern/diff/D34938125)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938125).

Commit:
Update on "Add api for changing function executor settings, hook up execution with decomposition registry"


Make the execution settings mutable on function_impl so that we can set it for running op decompositions. Add mapping to function objects and show example in test of executing op decompositions.

Differential Revision: [D34938125](https://our.internmc.facebook.com/intern/diff/D34938125)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938125).
